File: packages/on_audio_query/on_audio_query_ios/ios/Classes/methods/observers/GenresObserver.swift
import MediaPlayer

class GenresObserver : NSObject, FlutterStreamHandler {
    
    // Main parameters
    private var query: GenresQuery?
    private var sink: FlutterEventSink?
    private let library = MPMediaLibrary.default()
    private let notification = NotificationCenter.default
    
    // [Internal] variable to detect when the observer is running or not.
    private var pIsRunning: Bool = false
    // [Get] variable to detect when the observer is running or not.
    var isRunning: Bool {
        get { pIsRunning }
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        // Define the [sink].
        // The sink is used to define/send a event(change) or error.
        sink = events
        
        // Setup the 'query' class.
        query = GenresQuery(
            sink: self.sink,
            args: arguments as? [String: Any]
        )
        
        // If 'observe' method is called and the observer is already running
        // don't 'init' another notification observer.
        if !pIsRunning {
            // 'Tell' to library send notification when changing.
            library.beginGeneratingLibraryChangeNotifications()
            
            // Will be 'called' everytime the MPMediaLibrary change.
            // TODO: Currently will 'fire' any change. We need detect only 'genres' change.
            notification.addObserver(forName: .MPMediaLibraryDidChange, object: nil, queue: nil, using: { _ in
                self.query?.queryGenres()
            })
        }
        
        // Define the [ArtistsObserver] as running.
        pIsRunning = true
        
        // Send the initial data.
        query?.queryGenres()
        
        // No errors.
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        // Define [isRunning] as false.
        pIsRunning = false
        
        // 'Tell' to library not send notification when changing.
        library.endGeneratingLibraryChangeNotifications()
        
        // Stop listening the [MPMediaLibrary].
        notification.removeObserver(
            self,
            name: .MPMediaLibraryDidChange,
            object: nil
        )
        
        // 'Cancel' the [sink] and [query].
        sink = nil
        query = nil
        
        // No erros.
        return nil
    }
}

File: packages/on_audio_query/on_audio_query_android/android/src/main/kotlin/com/lucasjosino/on_audio_query/methods/observers/GenresObserver.kt
package com.lucasjosino.on_audio_query.methods.observers

import android.content.Context
import android.database.ContentObserver
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.provider.MediaStore
import com.lucasjosino.on_audio_query.methods.queries.GenresQuery
import io.flutter.plugin.common.EventChannel

class GenresObserver(
    private val context: Context,
) : ContentObserver(Handler(Looper.getMainLooper())), EventChannel.StreamHandler {

    companion object {
        // Fixed URI used as a path to [Genres].
        // Every 'observer' has your own URI.
        private val URI: Uri = MediaStore.Audio.Genres.EXTERNAL_CONTENT_URI
    }

    // [Get] variable to detect when the observer is running or not.
    val isRunning: Boolean get() = pIsRunning

    // Main parameters
    private val query: GenresQuery get() = GenresQuery()

    private var sink: EventChannel.EventSink? = null
    private var args: Map<*, *>? = null

    // [Internal] variable to detect when the observer is running or not.
    private var pIsRunning: Boolean = false

    // This function will be 'called' everytime the MediaStore -> Genres change.
    override fun onChange(selfChange: Boolean) {
        // We reuse the [queryGenres] from [GenresQuery] and 'query' everytime some change is detected.
        query.init(context, sink = sink, args = args)
    }

    // This function will be 'called' everytime the Flutter [EventChannel] is called.
    // [GenresObserver] event channel name: 'com.lucasjosino.on_audio_query/genres_observer'.
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        // Define the [sink] and [args]
        // The sink is used to define/send a event(change) or error.
        sink = events
        args = arguments as Map<*, *>

        // Register this class to observe the [MediaStore].
        context.contentResolver.registerContentObserver(URI, true, this)

        // Define the [GenresObserver] as running.
        pIsRunning = true

        // Send the initial data.
        query.init(context, sink = sink, args = args)
    }

    // This function will cancel the listener between the Dart <-> Native communication and at the
    // same time will unregister the 'MediaStore' listener.
    override fun onCancel(arguments: Any?) {
        // Stop listening the [MediaStore]
        context.contentResolver.unregisterContentObserver(this)

        // Cancel the [sink] and define [isRunning] as false.
        sink = null
        pIsRunning = false
    }
}